### ISSUE TITLE
Install testnet but ION generate mainnet DID

### ISSUE THREAD

**Issue by napoleon211092** [decentralized-identity/ion#227]
Hi,
I have success installed ION testnet.
But when I use command "ion operation create", it returned "did:ion:EiCq6M1QSsDs-_gf3ZgHyvd4L7-KxWEjCQWq8ee03VzN8Q". That DID belongs to mainet. 
Please help me.
Figure 1: MongoDB storage
![image](https://user-images.githubusercontent.com/8719020/131252364-e83128c4-c705-4c43-8af4-cc6b7bcbf830.png)
Figure 2: create DID
![image](https://user-images.githubusercontent.com/8719020/131252170-262ec1a2-b646-4c3b-b6a7-9bca328f0de9.png)
Note: 
I use:
-ION lastest (version 1.03)
-Bitcoin version bitcoin-0.20.1
Here is my testnet configuration:
1. testnet-bitcoin-config.json
{
  "bitcoinDataDirectory": "/root/.bitcoin/testnet3",
  "bitcoinFeeSpendingCutoffPeriodInBlocks": 1,
  "bitcoinFeeSpendingCutoff": 0.001,
  "bitcoinPeerUri": "http://localhost:18332",
  "bitcoinRpcUsername": "admin",
  "bitcoinRpcPassword": "admin",
  "bitcoinWalletOrImportString": "cRkQMNJkXuFvrgQ1QLWztsZXsucNQV7v6qrXGwT1P1RHZFkvm6xU",
  "databaseName": "ion-testnet-bitcoin",
  "genesisBlockNumber": 1900000,
  "logRequestError": true,
  "mongoDbConnectionString": "mongodb://127.0.0.1:27017/",
  "port": 3002,
  "sidetreeTransactionFeeMarkupPercentage": 1,
  "sidetreeTransactionPrefix": "ion:",
  "transactionPollPeriodInSeconds": 60,
  "valueTimeLockUpdateEnabled": false,
  "valueTimeLockAmountInBitcoins": 0,
  "valueTimeLockPollPeriodInSeconds": 600,
  "valueTimeLockTransactionFeesAmountInBitcoins": 0.0001
}
2. testnet-core-config.json
{
  "batchingIntervalInSeconds": 600,
  "blockchainServiceUri": "http://127.0.0.1:3002",
  "databaseName": "ion-testnet-core",
  "didMethodName": "ion:test",
  "ipfsHttpApiEndpointUri": "http://127.0.0.1:5001",
  "maxConcurrentDownloads": 20,
  "mongoDbConnectionString": "mongodb://127.0.0.1:27017/",
  "observingIntervalInSeconds": 60,
  "port": 3000
}
Thank you in advance.
P/s: 
I have tested with exist DID for testnet in "Section 8  Verify ION is working properly" on this link "https://identity.foundation/ion/install-guide/"
testnet: http://localhost:3000/identifiers/did:ion:test:EiClWZ1MnE8PHjH6y4e4nCKgtKnI1DK1foZiP61I86b6pw
And it worked with my local node
![image](https://user-images.githubusercontent.com/8719020/131255727-8cfe1f65-2a8f-4373-b5e0-11eb894ff3db.png)
But with my DID (generate by myself), it shows this error
![image](https://user-images.githubusercontent.com/8719020/131256237-c95de106-6c69-43aa-904e-a2c058e42a2d.png)

I have tried "ION Tool" too (link https://github.com/decentralized-identity/ion-tools)
Step 1: I deploy ION Tool and use webgui generate test DID
![image](https://user-images.githubusercontent.com/8719020/131277442-a8967552-7c46-4c44-bc2e-a9af7346829f.png)
It returned "Create DID request object"
{
  "type": "create",
  "suffixData": {
    "deltaHash": "EiBkinyW-325gkM5LMUuTbtaTHmIcdC1GRy-w52fZLxMIg",
    "recoveryCommitment": "EiA8KxRI9DDcRIoicG3vgQID7zt2Z6ziNxROc8ZJokW8Zg"
  },
  "delta": {
    "updateCommitment": "EiBc-U9p3oha3EOJPxpBVrwPyl4ANg3amD27rzwmZ1TsYg",
    "patches": [
      {
        "action": "replace",
        "document": {
          "publicKeys": [
            {
              "id": "key-1",
              "type": "EcdsaSecp256k1VerificationKey2019",
              "publicKeyJwk": {
                "kty": "EC",
                "crv": "secp256k1",
                "x": "KyEpPuK82-Ngi2JZYQGBM8fvRDlsMW6BoL30me1fzGY",
                "y": "3Zx-TFk3bYWVhBEpX8XhanMkjgbrQRWx8xh0TIEzp8s"
              },
              "purposes": [
                "authentication"
              ]
            }
          ],
          "services": [
            {
              "id": "domain-1",
              "type": "LinkedDomains",
              "serviceEndpoint": "13.251.63.234:3000"
            }
          ]
        }
      }
    ]
  }
}
Step 2: I use this object to create ION DID by call ION API
![image](https://user-images.githubusercontent.com/8719020/131277619-6fd59f98-f42e-4b38-82ca-b6ff81b67c04.png)
And it returned this:
{
    "@context": "https://w3id.org/did-resolution/v1",
    "didDocument": {
        "id": "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA",
        "@context": [
            "https://www.w3.org/ns/did/v1",
            {
                "@base": "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA"
            }
        ],
        "service": [
            {
                "id": "#domain-1",
                "type": "LinkedDomains",
                "serviceEndpoint": "13.251.63.234:3000"
            }
        ],
        "verificationMethod": [
            {
                "id": "#key-1",
                "controller": "",
                "type": "EcdsaSecp256k1VerificationKey2019",
                "publicKeyJwk": {
                    "kty": "EC",
                    "crv": "secp256k1",
                    "x": "KyEpPuK82-Ngi2JZYQGBM8fvRDlsMW6BoL30me1fzGY",
                    "y": "3Zx-TFk3bYWVhBEpX8XhanMkjgbrQRWx8xh0TIEzp8s"
                }
            }
        ],
        "authentication": [
            "#key-1"
        ]
    },
    "didDocumentMetadata": {
        "method": {
            "published": false,
            "recoveryCommitment": "EiA8KxRI9DDcRIoicG3vgQID7zt2Z6ziNxROc8ZJokW8Zg",
            "updateCommitment": "EiBc-U9p3oha3EOJPxpBVrwPyl4ANg3amD27rzwmZ1TsYg"
        },
        "canonicalId": "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA"
    }
}
Step 3: I resolve with DID "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA", but It proved "not found"
![image](https://user-images.githubusercontent.com/8719020/131277757-ecb1b63d-e33e-4fe0-8e6c-dd96d218a3a8.png)
In conclusion, I sumarize my problem:
1. If I use command "ion operation create" to generate DID, it will return DID for mainet depite I deploy testnet.
2. If I use "ION DID tool" to generate DID, it return DID for testnet (right) but I can't resolve that DID.
Please help me.

**Comment by thehenrytsai** [comment]
Hi @napoleon211092, thanks for the thorough description. Let me try to an answer your questions, please do not hesitate to let me know if I missed anything.

> But when I use command "ion operation create"...

The CLI is completely experimental (as it is using a test library) and on development pause since last year, we did not anticipate anyone would know about it since we didn't advertise it on front page, but I have just opened a PR to make it use the ION SDK instead, it should be merged soon, I am not sure if it was working before my PR (probably was), but it should definitely work now.

> That DID belongs to mainet.

The CLI was written with the assumption that it targets mainnet. But it is fine, the request will work perfectly fine against the testnet. The `test:` prefix in the DID string is purely cosmetic for the most part.

> But with my DID (generate by myself), it shows this error

Since you are testing testnet DIDs and your node is setup as such (`"didMethodName": "ion:test",` seen in your config above), you must therefore prefix your unique DID suffix string with `did:ion:test:<unique_suffix>`, you didn't in your screenshot, hence the error. I hope this makes sense. The error message you are seeing in the screenshot also says exactly that.

> Step 3: I resolve with DID "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA", but It proved "not found"

ION DID tool should work. Sanity: assuming you've setup the node to be able to write, did you wait for that transaction to be confirmed? You can check by checking the transaction hash that is printed out when batch writer kicks in every 10 minutes (based on your conifg `"batchingIntervalInSeconds": 600,`). Even though the post request immediately returns a DID Document, that is just for your convenience to know that 1. the node accepted your request and queues it, and 2. shows what your DID Document will look like once it is confirmed, purely for your convenience. But we all know bitcoin transactions takes a while to confirm, especially on testnet, where the confirmation time is less predictable.

**Comment by napoleon211092** [comment]
> 
> 
> Hi @napoleon211092, thanks for the thorough description. Let me try to an answer your questions, please do not hesitate to let me know if I missed anything.
> 
> > But when I use command "ion operation create"...
> 
> The CLI is completely experimental (as it is using a test library) and on development pause since last year, we did not anticipate anyone would know about it since we didn't advertise it on front page, but I have just opened a PR to make it use the ION SDK instead, it should be merged soon, I am not sure if it was working before my PR (probably was), but it should definitely work now.
> 
> > That DID belongs to mainet.
> 
> The CLI was written with the assumption that it targets mainnet. But it is fine, the request will work perfectly fine against the testnet. The `test:` prefix in the DID string is purely cosmetic for the most part.
> 
> > But with my DID (generate by myself), it shows this error
> 
> Since you are testing testnet DIDs and your node is setup as such (`"didMethodName": "ion:test",` seen in your config above), you must therefore prefix your unique DID suffix string with `did:ion:test:<unique_suffix>`, you didn't in your screenshot, hence the error. I hope this makes sense. The error message you are seeing in the screenshot also says exactly that.
> 
> > Step 3: I resolve with DID "did:ion:test:EiC3YoSodQ20tJcgKjLXr65BHr2KwnQWsUnm3VOiYUFMeA", but It proved "not found"
> 
> ION DID tool should work. Sanity: assuming you've setup the node to be able to write, did you wait for that transaction to be confirmed? You can check by checking the transaction hash that is printed out when batch writer kicks in every 10 minutes (based on your conifg `"batchingIntervalInSeconds": 600,`). Even though the post request immediately returns a DID Document, that is just for your convenience to know that 1. the node accepted your request and queues it, and 2. shows what your DID Document will look like once it is confirmed, purely for your convenience. But we all know bitcoin transactions takes a while to confirm, especially on testnet, where the confirmation time is less predictable.

Thank your for your answer Mr @thehenrytsai.
I have tried ION Tool and write an REST API for this.
https://github.com/thanhlam2110/ION-DID-API

**Comment by thehenrytsai** [comment]
Thanks @napoleon211092, closing this issue. please feel free to reopen or open new issue as needed.
